### ISSUE TITLE
Inserting Newlines in Diff and Incorrectly Comparing

### ISSUE THREAD

**Issue by scottmk** [DougTy/sublime-compare-side-by-side#45]
I don't know how to properly describe this issue, so I'm just going to provide a screenshot.

Pay particular attention to LL9-23.

These documents have the same number of lines, but whenever this plugin determines that there is a conflict, it inserts newlines into one of the documents, which makes it incredibly difficult to compare.

It also is detecting conflicts that make no sense, like L22 on the red side, which seems to be saying that the P, r, and d in "Prod" differ from the green side, which is obviously not true.

Using other diff tools yielded the behavior I expected, so I can only assume this is a bug.

![screen shot 2018-08-03 at 16 43 05](https://user-images.githubusercontent.com/3020184/43672658-a8546bc2-9767-11e8-95d8-2bffbc8fbf3c.png)

**Comment by DougTy** [comment]
I agree, that doesn't look quite right.1. What OS & build of Sublime are you on? Are you using the latest version of the plugin (from package control?)2. Could it a whitespace issue? Try enabling `ignore_whitespace` in the plugin settings.3. Can you paste some text I may use to test this myself? Just the visible portion of the screenshot will do fine if you're okay with that.On Aug 3, 2018 10:53 PM, "Michael K. Scott" <notifications@github.com> wrote:I don't know how to properly describe this issue, so I'm just going to provide a screenshot.
Pay particular attention to LL9-23.
These documents have the same number of lines, but whenever this plugin determines that there is a conflict, it inserts newlines into one of the documents, which makes it incredibly difficult to compare.
It also is detecting conflicts that make no sense, like L22 on the red side, which seems to be saying that the P, r, and d in "Prod" differ from the green side, which is obviously not true.
Using other diff tools yielded the behavior I expected, so I can only assume this is a bug.

**Comment by hadrielk** [comment]
This is what I get, which looks right:
<img width="1426" alt="screen shot 2018-08-04 at 4 50 55 pm" src="https://user-images.githubusercontent.com/3319094/43680352-ca8dfda8-9806-11e8-8d85-86a54acf4219.png">

**Comment by scottmk** [comment]
1. Mac OS X, latest version. I am also using the latest version of the plugin, which I downloaded from package control.
2. I suppose it could be, though to me they appear to have the same amount of whitespace. Moreover, it seems weird that only the 'o' in 'Prod' on L22 would be changed. It seems like the whole line should be highlighted at that point.
3. I just saw this, so I don't know if I still have the text, but if I do, I will paste it tomorrow. Sorry, I should have done that when I posted the issue.

**Comment by hadrielk** [comment]
> Moreover, it seems weird that only the 'o' in 'Prod' on L22 would be changed. It seems like the whole line should be highlighted at that point.

Ultimately this plugin uses Python's [difflib](https://docs.python.org/2/library/difflib.html) library to actually determine the differences between text. It's possible the sublime plugin is misinterpreting the results from difflib, but difflib must be detecting *some* differences for each line. We'll definitely need the original text to figure it out.

**Comment by scottmk** [comment]
[file2.txt](https://github.com/DougTy/sublime-compare-side-by-side/files/2279102/file2.txt)
[file1.txt](https://github.com/DougTy/sublime-compare-side-by-side/files/2279103/file1.txt)

**Comment by hadrielk** [comment]
I tried your files, and I don't get the difference you got. Instead I get 486 lines shown removed from file 1, which seems correct:

<img width="1436" alt="screen shot 2018-08-12 at 10 50 48 pm" src="https://user-images.githubusercontent.com/3319094/44010757-4ffa26fc-9e82-11e8-94de-a591c5105fc9.png">

**Comment by hadrielk** [comment]
Oh wait, I got it. I can get something like you showed - I forgot I use different settings than the default.
For some reason, the default `"ignore_whitespace": false` setting makes `difflib` generate a very different output for these files. (I have `"ignore_whitespace": true` for my user setting instead)

Go to `Sublime Text->Preferences->Package Settings->Compare Side-by-Side->Settings-User`, and put this in the opened file:

```
{
    "ignore_whitespace": true
}
```

**Comment by DougTy** [comment]
Thanks for debugging @hadrielk, I've been preoccupied with a family emergency the past week & really appreciate the help. I tried poking around with this for a couple hours yesterday but couldn't figure anything out, so I'm glad you could give us a lead.

I think I'll just make `"ignore_whitespace": true` the default, as from some quick testing this seems to be the default for other diffing tools as well (ie. in VSCode). Unless anyone can think of a reason why not, I think this is a good default to have.

As for the actual underlying problem, I truly have no idea why this is happening. If you try doing a manual replace-all on whitespace characters to remove them (within Sublime), you'll get the same result. It's as if there are some hidden characters that are messing with the diff, but I can't identify them, so maybe not. [The `ignore_whitespace` option is simply doing a regex replace for tabs and spaces to remove them](https://github.com/DougTy/sublime-compare-side-by-side/blob/master/sbs_compare.py#L315
), so I'm not sure why this would work in code but not manually.

**Comment by DougTy** [comment]
@scottmk Can you confirm if setting `"ignore_whitespace": true` resolves this for you?

I'll make this the new default for now, but keep the issue open to continue trying to figure this out & fix it later as something is still clearly not right while this is disabled.
